source/common/upstream/health_checker_impl.cc

```diff
@@ -145,6 +146,35 @@ void HealthCheckerImplBase::runCallbacks(HostSharedPtr host, bool changed_state)
   }
 }
 
+void HealthCheckerImplBase::HealthCheckerSinkImpl::setUnhealthy() {
+  // This is called cross thread. The cluster/health checker might already be gone.
+  std::shared_ptr<HealthCheckerImplBase> health_checker = health_checker_.lock();
+  if (health_checker) {
+    health_checker->setUnhealthyCrossThread(host_.lock());
+  }
+}
+
+void HealthCheckerImplBase::setUnhealthyCrossThread(const HostSharedPtr& host) {
+  // The threading here is complex. The cluster owns the only strong reference to the health
+  // checker. It might go away when we post to the main thread. We capture a weak reference and
+  // make sure it is still valid when we get to the other thread. Additionally, the host/session
```

Comment:
Would be helpful to be clearer here on which thread is which when referring to "other thread". I.e. we have a worker thread (presumably monitoring the passive header check) communicating with the main thread?